core/tee/tee_svc_storage.c

```diff
@@ -1322,9 +1338,34 @@ TEE_Result syscall_storage_obj_write(unsigned long obj, void *data, size_t len)
 		goto exit;
 
 	o->info.dataPosition += len;
-	if (o->info.dataPosition > o->info.dataSize)
+	if (o->info.dataPosition > o->info.dataSize) {
+		struct tee_svc_storage_head *head = NULL;
+		tee_fs_off_t off;
+
 		o->info.dataSize = o->info.dataPosition;
 
+		/* Get current offset */
+		off = o->pobj->fops->lseek(&res, o->fd, 0, TEE_FS_SEEK_CUR);
+
+		head = o->pobj->head;
+
+		if (head && head->magic == TEE_SVC_STORAGE_MAGIC) {
+			/* Update size */
+			head->ds_size = o->info.dataSize;
+			/* Update head into storage */
+			o->pobj->fops->lseek(&res, o->fd, 0, TEE_FS_SEEK_SET);
+
+			if (o->pobj->fops->write(&res, o->fd, head,
+					sizeof(*head)) != sizeof(*head)) {
+				EMSG("Failed to update head\n");
+				goto exit;
```

Comment:
Indeed `goto exit` may not be the best exit.

Would stored object be corrupted if payload is updated but not the head?